Fix displacy span stacking

rmitsch opened the pull request:
<!--- Provide a general summary of your changes in the title. -->

## Description
<!--- Use this section to describe your changes. If your changes required
testing, include information about the testing environment and the tests you
ran. If your test fixes a bug reported in an issue, don't forget to include the
issue number. If your PR is still a work in progress, that's totally fine – just
include a note to let us know. -->
Fix displacy span stacking by using the stack's last span's level instead of a counter not taking previous spans' levels into account.
Fixes https://github.com/explosion/spaCy/issues/13056.

Before: 
![image](https://github.com/explosion/spaCy/assets/7152225/d828b9a4-573e-4e03-9387-ac23d222a83d)

After:
![image](https://github.com/explosion/spaCy/assets/7152225/0ffa28f8-dba3-4fc8-8b8d-6d522ec62b6a)


### Types of change
<!-- What type of change does your PR cover? Is it a bug fix, an enhancement
or new feature, or a change to the documentation? -->
Bug fix

## Checklist
<!--- Before you submit the PR, go over this checklist and make sure you can
tick off all the boxes. [] -> [x] -->
- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

svlandeg reviewed (commented):
Is there a unit test we could write to double check this behaviour?

rmitsch commented:
> Is there a unit test we could write to double check this behaviour?

Added one. Also refactored the relevant functionality to allow checking the data prior to rendering it as HTML.
